### PR TITLE
Fix V8 API memory management and implement more APIs

### PR DESCRIPTION
**Pull request by 190n**
### What does this PR do?

- New V8 APIs (so far):
    - `v8::Context::GetIsolate`
    - `v8::String::NewFromOneByte`
- Make `HandleScope` growable (while ensuring pointer stability)
- Fix memory management in `HandleScopeBuffer`, `InternalFieldObject`, and `FunctionTemplate` so that V8 objects are not garbage collected, and add tests that segfault if those cases are broken
- Add more `static_assert`s to ensure V8 layout is matched where it matters
- Use `WTF::FixedVector` for storage in `InternalFieldObject` (should have less overhead as it is not resizable)
- Support `v8::Global` with tests
- Refactor the V8 tests to be simpler: instead of manually logging messages in the native module indicating a failed assertion, the expected value, and the actual value, I just print out the expression and its value. If the output differs from the same test run under Node.js, Bun's test runner will show a diff.
    - Update the tests from #13417 with this format

### How did you verify your code works?

- [x] I included a test for the new code, or existing tests cover it
- [x] I ran my tests locally and they pass (`bun-debug test test-file-name.test`)

### CODE REVIEW

**Comment by robobun** [comment]
@190n, your commit 72bd3770a6c133ec82b8f4dd830d1c61b5bc39d0 is building: [`#2101`](https://buildkite.com/bun/bun/builds/2101)

File: src/bun.js/bindings/v8/V8FunctionTemplate.cpp [hunk 1]
```diff
@@ -54,7 +54,7 @@ Local<FunctionTemplate> FunctionTemplate::New(
         vm, structure, callback, jsc_data);
     functionTemplate->finishCreation(vm);
 
-    return isolate->currentHandleScope()->createLocal<FunctionTemplate>(functionTemplate);
+    return isolate->currentHandleScope()->createLocal<FunctionTemplate>(isolate->vm(), functionTemplate);
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/v8/V8FunctionTemplate.cpp]:
```suggestion
    return isolate->currentHandleScope()->createLocal<FunctionTemplate>(vm, functionTemplate);
```

File: src/bun.js/bindings/v8/V8ObjectTemplate.cpp [hunk 2]
```diff
@@ -47,7 +47,7 @@ Local<ObjectTemplate> ObjectTemplate::New(Isolate* isolate, Local<FunctionTempla
     auto* objectTemplate = new (NotNull, JSC::allocateCell<ObjectTemplate>(vm)) ObjectTemplate(vm, structure);
     // TODO pass constructor
     objectTemplate->finishCreation(vm);
-    return isolate->currentHandleScope()->createLocal<ObjectTemplate>(objectTemplate);
+    return isolate->currentHandleScope()->createLocal<ObjectTemplate>(isolate->vm(), objectTemplate);
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/v8/V8ObjectTemplate.cpp]:
```suggestion
    return isolate->currentHandleScope()->createLocal<ObjectTemplate>(vm, objectTemplate);
```

File: src/bun.js/bindings/v8/V8String.cpp [hunk 3]
```diff
@@ -28,7 +28,17 @@ MaybeLocal<String> String::NewFromUtf8(Isolate* isolate, char const* data, NewSt
     auto string = WTF::String::fromUTF8ReplacingInvalidSequences(span);
     RELEASE_ASSERT(!string.isNull());
     JSString* jsString = JSC::jsString(isolate->vm(), string);
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/v8/V8String.cpp]:
```suggestion
    auto vm = isolate->vm();
    JSString* jsString = JSC::jsString(vm, string);
```

File: src/bun.js/bindings/v8/V8String.cpp [hunk 4]
```diff
@@ -28,7 +28,17 @@ MaybeLocal<String> String::NewFromUtf8(Isolate* isolate, char const* data, NewSt
...
-    return MaybeLocal<String>(isolate->globalInternals()->currentHandleScope()->createLocal<String>(jsString));
+    return MaybeLocal<String>(isolate->currentHandleScope()->createLocal<String>(isolate->vm(), jsString));
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/v8/V8String.cpp]:
```suggestion
    return MaybeLocal<String>(isolate->currentHandleScope()->createLocal<String>(vm, jsString));
```

File: src/bun.js/bindings/v8/V8String.cpp [hunk 5]
```diff
@@ -28,7 +28,17 @@ MaybeLocal<String> String::NewFromUtf8(Isolate* isolate, char const* data, NewSt
...
+}
+
+MaybeLocal<String> String::NewFromOneByte(Isolate* isolate, const uint8_t* data, NewStringType type, int length)
+{
+    (void)type;
+    RELEASE_ASSERT(length < 0, "Non-inferred string length in v8::String::NewFromOneByte is not implemented");
+
+    auto string = WTF::String::fromLatin1(reinterpret_cast<const char*>(data));
+    JSString* jsString = JSC::jsString(isolate->vm(), string);
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/v8/V8String.cpp]:
```suggestion
    auto& vm = isolate->vm();
    JSString* jsString = JSC::jsString(vm, string);
```

File: src/bun.js/bindings/v8/V8Handle.h [hunk 6]
```diff
@@ -4,10 +4,53 @@
...
+        // using a smi value for map is most likely to catch bugs as almost every access will expect
+        // map to be a pointer (and even if the assertion is bypassed, it'll be a null pointer)
+        : tagged_map(0)
+        , contents({ .raw = nullptr })
+    {
+    }
+
+    ObjectLayout(const Map* map_ptr, JSC::JSCell* cell, JSC::VM& vm, const JSC::JSCell* owner)
+        : tagged_map(const_cast<Map*>(map_ptr))
+        , contents({ .cell = JSC::WriteBarrier<JSC::JSCell>(vm, owner, cell) })
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/v8/V8Handle.h:28]: Do we need a destructor to call `contents.cell.clear()` since its a union?

Comment (190n) [reply src/bun.js/bindings/v8/V8Handle.h:28]: Is that required to be called? I don't see destructors in WriteBarrier or WriteBarrierBase that would be skipped

**Pull request merged**